File: 2024/Python/Challenge#5.py
from typing import List, TypedDict, Literal


class Shoe(TypedDict):
    type: Literal['I', 'R']
    size: int

# def organize_shoes(shoes: List[Shoe]) -> List[int]:
#     shoes_by_size = {}
#     for shoe in shoes:
#         _type = shoe['type']
#         size = shoe['size']
        
#         if size not in shoes_by_size:
#             shoes_by_size[size] = {"I": 0, "R": 0}
    
#         shoes_by_size[size][_type] += 1
    
#     available_shoes = []
#     for size, types in shoes_by_size.items():
#         I = types.get('I', 0)
#         R = types.get('R', 0)
#         available_shoes += [size] * min(I, R)  # Version sin min() --> available_shoes += [size] * (I if I < R else R)
    
#     return available_shoes

def organize_shoes(shoes: List[Shoe]) -> List[int]:
    available_shoes = []
    shoes_by_size = {}

    for shoe in shoes:
        size = shoe["size"]
        shoe_type = shoe["type"]
        opposite_type = "R" if shoe_type == "I" else "I"

        if shoes_by_size.get(size, {}).get(opposite_type, 0) > 0:
            available_shoes.append(size)
            shoes_by_size[size][opposite_type] -= 1
        else:
            if size not in shoes_by_size:
                shoes_by_size[size] = {}
            shoes_by_size[size][shoe_type] = shoes_by_size[size].get(shoe_type, 0) + 1

    return available_shoes


shoes = [
  { 'type': 'I', 'size': 38 },
  { 'type': 'R', 'size': 38 },
  { 'type': 'R', 'size': 42 },
  { 'type': 'I', 'size': 41 },
  { 'type': 'I', 'size': 42 }
]
pairs = organize_shoes(shoes)
print(pairs)  # [38, 42]

shoes2 = [
  { 'type': 'I', 'size': 38 },
  { 'type': 'R', 'size': 38 },
  { 'type': 'I', 'size': 38 },
  { 'type': 'I', 'size': 38 },
  { 'type': 'R', 'size': 38 }
]
pairs2 = organize_shoes(shoes2)
print(pairs2)  # [38, 38]

shoes3 = [
  { 'type': 'I', 'size': 38 },
  { 'type': 'R', 'size': 36 },
  { 'type': 'R', 'size': 42 },
  { 'type': 'I', 'size': 41 },
  { 'type': 'I', 'size': 43 }
]
pairs3 = organize_shoes(shoes3)
print(pairs3)  # []

File: 2024/TypeScript/Challenge#5.ts
type Shoe = {
    type: 'I' | 'R'
    size: number
}

// function organizeShoes(shoes: Shoe[]): number[] {
//     const shoesBySize = new Map<number, Map<string, number>>();

//     for (const shoe of shoes) {
//         if (!shoesBySize.has(shoe.size)) {
//             shoesBySize.set(shoe.size, new Map<string, number>());
//         }

//         const typeCount = shoesBySize.get(shoe.size);
//         typeCount?.set(shoe.type, (typeCount?.get(shoe.type) || 0) + 1);
//     }

//     const availableShoes = new Array<number>();
//     for (const [size, types] of shoesBySize) {
//         const I: number = types.get('I') || 0;
//         const R: number = types.get('R') || 0;
//         availableShoes.push(...Array(Math.min(I, R)).fill(size));
//     }

//     return availableShoes;
// }

function organizeShoes(shoes: Shoe[]): number[] {
    const availableShoes: number[] = [];
    const shoesBySize = new Map<number, Map<string, number>>();

    for (const shoe of shoes) {
        const { size, type } = shoe;
        const oppositeType = type === "I" ? "R" : "I";

        if (!shoesBySize.has(size)) {
            shoesBySize.set(size, new Map<string, number>());
        }

        const typeCount = shoesBySize.get(size)!;
        
        if (typeCount.get(oppositeType)! > 0) {
            availableShoes.push(size);
            typeCount.set(oppositeType, typeCount.get(oppositeType)! - 1);
        } else {
            typeCount.set(type, (typeCount.get(type) || 0) + 1);
        }
    }

    return availableShoes;
}


const shoes: Shoe[] = [
    { type: 'I', size: 38 },
    { type: 'R', size: 38 },
    { type: 'R', size: 42 },
    { type: 'I', size: 41 },
    { type: 'I', size: 42 }
]
const pairs = organizeShoes(shoes)
console.log(pairs)  // [38, 42]

const shoes2: Shoe[] = [
    { type: 'I', size: 38 },
    { type: 'R', size: 38 },
    { type: 'I', size: 38 },
    { type: 'I', size: 38 },
    { type: 'R', size: 38 }
]
const pairs2 = organizeShoes(shoes2)
console.log(pairs2)  // [38, 38]

const shoes3: Shoe[] = [
    { type: 'I', size: 38 },
    { type: 'R', size: 36 },
    { type: 'R', size: 42 },
    { type: 'I', size: 41 },
    { type: 'I', size: 43 }
]
const pairs3 = organizeShoes(shoes3)
console.log(pairs3)  // []
